hedera-node/src/main/java/com/hedera/services/stream/RecordStreamFileWriter.java

```diff
@@ -0,0 +1,490 @@
+	 */
+	@Override
+	public void clear() {
+		if (dosMeta != null) {
+			try {
+				dosMeta.close();
+				metadataStreamDigest.reset();
+				dosMeta = null;
+			} catch (IOException e) {
+				LOG.warn(EXCEPTION.getMarker(), "RecordStreamFileWriter::clear Exception in closing dosMeta", e);
+			}
+		}
+		recordStreamFileBuilder = null;
+		LOG.debug(OBJECT_STREAM.getMarker(), "RecordStreamFileWriter::clear executed.");
+	}
+
+	public void close() {
+		this.closeCurrentAndSign();
+		LOG.debug(LogMarker.FREEZE.getMarker(), "RecordStreamFileWriter finished writing the last object, is stopped");
+	}
+
+	public void setStartWriteAtCompleteWindow(final boolean startWriteAtCompleteWindow) {
+		this.startWriteAtCompleteWindow = startWriteAtCompleteWindow;
+		LOG.debug(OBJECT_STREAM.getMarker(),
+				"RecordStreamFileWriter::setStartWriteAtCompleteWindow: {}", startWriteAtCompleteWindow);
+	}
+
+	public boolean getStartWriteAtCompleteWindow() {
+		return this.startWriteAtCompleteWindow;
+	}
+
+	/**
+	 * Helper method that serializes a RecordStreamFile.
+	 * Uses deterministic serialization to ensure multiple invocations on the same
+	 * object lead to identical serialization.
+	 * @param recordStreamFileProtoBuilder
+	 * 		the RecordStreamFile object that needs to be serialized
+	 * @return
+	 * 		the serialized bytes
+	 */
+	private byte[] serialize(final RecordStreamFile.Builder recordStreamFileProtoBuilder) throws IOException {
+		final var recordStreamFileProto = recordStreamFileProtoBuilder.build();
+		final var result = new byte[recordStreamFileProto.getSerializedSize()];
+		final var output = CodedOutputStream.newInstance(result);
+		output.useDeterministicSerialization();
+		recordStreamFileProto.writeTo(output);
+		output.checkNoSpaceLeft();
+		return result;
+	}
+
+	private HashObject toProto(final Hash hash) {
+		return HashObject.newBuilder()
+				.setAlgorithm(HashAlgorithm.SHA_384)
+				.setLength(hash.getDigestType().digestLength())
+				.setHash(ByteString.copyFrom(hash.getValue()))
+				.build();
+	}
+
+	private void createSignatureFile(final File relatedRecordStreamFile) {
+		// get entire hash of current record stream file
+		final var entireHash = new Hash(streamDigest.digest(), DigestType.SHA_384);
+		// get metaData hash of current record stream file
+		final var metaHash = new Hash(metadataStreamDigest.digest(), DigestType.SHA_384);
+
+		// create proto messages for signature file
+		final var fileSignature = generateSignatureObject(entireHash);
+		final var metadataSignature = generateSignatureObject(metaHash);
+		final var signatureFile = SignatureFile.newBuilder()
+				.setFileSignature(fileSignature)
+				.setMetadataSignature(metadataSignature);
+
+		// create signature file
+		final var sigFilePath = generateSigFilePath(relatedRecordStreamFile);
+		try (final var fos = new FileOutputStream(sigFilePath)) {
+			fos.write(recordFileVersion); // version in signature files is 1 byte, compared to 4 in record files
```

Comment:
In the future, record file version and signature version may not be the same. I think we should have a separate variable to track signature version and not link the two.